src/query/remote/client.go

```diff
@@ -55,40 +92,103 @@ type grpcClient struct {
 	pools       encoding.IteratorPools
 	poolErr     error
 	opts        m3db.Options
+	closeCh     chan struct{}
+	metrics     grpcClientMetrics
 }
 
-const initResultSize = 10
+type grpcClientMetrics struct {
+	healthCheckSuccess tally.Counter
+	healthCheckError   tally.Counter
+}
+
+func newGRPCClientMetrics(s tally.Scope) grpcClientMetrics {
+	s = s.SubScope("remote-client")
+	return grpcClientMetrics{
+		healthCheckSuccess: s.Tagged(map[string]string{
+			healthCheckMetricResultTag: "success",
+		}).Counter(healthCheckMetricName),
+		healthCheckError: s.Tagged(map[string]string{
+			healthCheckMetricResultTag: "error",
+		}).Counter(healthCheckMetricName),
+	}
+}
 
 // NewGRPCClient creates a new remote GRPC client.
 func NewGRPCClient(
+	name string,
 	addresses []string,
 	poolWrapper *pools.PoolWrapper,
 	opts m3db.Options,
+	instrumentOpts instrument.Options,
 	additionalDialOpts ...grpc.DialOption,
 ) (Client, error) {
 	if len(addresses) == 0 {
 		return nil, errors.ErrNoClientAddresses
 	}
 
+	// Set name if using a named client.
+	if remote := strings.TrimSpace(name); remote != "" {
+		instrumentOpts = instrumentOpts.
+			SetMetricsScope(instrumentOpts.MetricsScope().Tagged(map[string]string{
+				"remote-name": remote,
+			}))
+	}
+
+	scope := instrumentOpts.MetricsScope()
+	interceptorOpts := xgrpc.InterceptorInstrumentOptions{Scope: scope}
+
 	resolver := newStaticResolver(addresses)
 	balancer := grpc.RoundRobin(resolver)
-	dialOptions := []grpc.DialOption{
+	dialOptions := append([]grpc.DialOption{
 		grpc.WithBalancer(balancer),
 		grpc.WithInsecure(),
-	}
+		grpc.WithUnaryInterceptor(xgrpc.UnaryClientInterceptor(interceptorOpts)),
+		grpc.WithStreamInterceptor(xgrpc.StreamClientInterceptor(interceptorOpts)),
+	}, defaultDialOptions...)
 	dialOptions = append(dialOptions, additionalDialOpts...)
 	cc, err := grpc.Dial("", dialOptions...)
 	if err != nil {
 		return nil, err
 	}
 
 	client := rpc.NewQueryClient(cc)
-	return &grpcClient{
+	c := &grpcClient{
 		client:      client,
 		connection:  cc,
 		poolWrapper: poolWrapper,
 		opts:        opts,
-	}, nil
+		closeCh:     make(chan struct{}),
+		metrics:     newGRPCClientMetrics(scope),
+	}
+	go c.healthCheckUntilClosed()
+	return c, nil
+}
+
+func (c *grpcClient) healthCheckUntilClosed() {
+	ticker := time.NewTicker(healthCheckInterval)
+	defer ticker.Stop()
+
+	req := &rpc.HealthRequest{}
+	for {
+		// Perform immediately so first check isn't delayed.
+		ctx, cancel := context.WithTimeout(context.Background(),
+			healthCheckTimeout)
+		_, err := c.client.Health(ctx, req)
+		cancel()
+		if err != nil {
+			c.metrics.healthCheckError.Inc(1)
```

Comment:
would some kind of rate limiting logging help with debugging the issue?

Reply:
Possibly, I can add some debug logs perhaps here, that way can at least temporarily turn it on.

Zap logger can actually dynamically change the level of logging, we just don't offer a way to do that yet (i.e. endpoint), we really should: https://github.com/uber-go/zap/issues/556.